ambry-commons/src/main/java/com.github.ambry.commons/HelixNotifier.java

```diff
@@ -0,0 +1,209 @@
+    }
+    String topicPath = makeTopicPath(topic);
+    HelixPropertyListener helixListener = new HelixPropertyListener() {
+      @Override
+      public void onDataChange(String path) {
+        logger.trace("Message is changed for topic {} at path {}", topic, path);
+        sendMessage(topicListener, topic, path);
+      }
+
+      @Override
+      public void onDataCreate(String path) {
+        logger.trace("Message is created for topic {} at path {}", topic, path);
+        sendMessage(topicListener, topic, path);
+      }
+
+      @Override
+      public void onDataDelete(String path) {
+        // for now, this is a no-op when a ZNRecord for a topic is deleted, since no
+        // topic deletion is currently supported.
+        logger.trace("Message is deleted for topic {} at path {}", topic, path);
+      }
+    };
+    topicListenerToHelixListenerMap.put(topicListener, helixListener);
+    List<TopicListener> topicListeners = topicToListenerMap.get(topic);
+    if (topicListeners == null) {
+      topicListeners = new ArrayList<>();
+    }
+    topicListeners.add(topicListener);
+    topicToListenerMap.put(topic, topicListeners);
+    helixStore.subscribe(topicPath, helixListener);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void unsubscribe(String topic, TopicListener<String> topicListener) {
+    if (topic == null) {
+      throw new IllegalArgumentException("topic cannot be null");
+    }
+    if (topicListener == null) {
+      throw new IllegalArgumentException("topicListener cannot be null");
+    }
+    String topicPath = makeTopicPath(topic);
+    HelixPropertyListener helixListener = topicListenerToHelixListenerMap.remove(topicListener);
+    helixStore.unsubscribe(topicPath, helixListener);
+    List<TopicListener> topicListeners = topicToListenerMap.get(topic);
+    if (topicListeners != null) {
+      topicListeners.remove(topicListener);
+    }
+  }
+
+  /**
+   * Gets all the {@link TopicListener}s for the specified topic.
+   * @param topic The topic to get its listeners.
+   * @return A list of {@link TopicListener}s that have subscribed to the topic.
+   */
+  public List<TopicListener> getListenersByTopic(String topic) {
+    if (topic == null || topicToListenerMap.get(topic) == null) {
+      return new ArrayList<>();
+    }
+    return Collections.unmodifiableList(topicToListenerMap.get(topic));
+  }
+
+  /**
+   * Sends message to {@link TopicListener}.
+   * @param topicListener The {@link TopicListener} to send the message.
+   * @param topic The topic which the message belongs to.
+   * @param path The path to the topic.
+   */
+  private void sendMessage(TopicListener topicListener, String topic, String path) {
+    String message = null;
+    try {
+      ZNRecord zNRecord = helixStore.get(path, null, AccessOption.PERSISTENT);
+      if (zNRecord != null) {
+        message = zNRecord.getSimpleField(MESSAGE_KEY);
+      }
+      topicListener.processMessage(topic, message);
```

Comment:
That's a good question, and I do not have a very solid answer to this. Originally I was thinking if this method gets called, there must be some change event happened on the corresponding zk path, so the notifier should try its best to deliver this change to its listeners. Before the final delivery, i.e., before calling `topicListener.onMessageReceive(topic, message)`, there could be additional things happen, so there is no guarantee the `ZNRecord`is still there.

However when reading your comments I gave a second thought on this, and feel it is not responsible to just deliver a `null` message to the listener. Logging might be a better option.